### PR TITLE
ASRC-46: calculate all probability outcomes

### PR DESCRIPTION
**Pull request by dgtombs**
The calculation code already supported calculating multiple outcomes, but this PR improves the outcome display. It also fixes an edge case where values for different variables with the same display name were collapsed, and it simplifies the missing values code.

### CODE REVIEW

File: srcalc/src/main/java/gov/va/med/srcalc/domain/model/ModelTerm.java [hunk 1]
```diff
@@ -6,9 +6,10 @@
 
 import com.google.common.collect.ImmutableSet;
 
+
+
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/domain/model/ModelTerm.java]: ??

File: srcalc/src/main/java/gov/va/med/srcalc/domain/model/SingleVariableTerm.java [hunk 2]
```diff
@@ -51,10 +48,7 @@ public float getSummand(final Map<Variable, Value> inputValues) throws MissingVa
         final Value value = inputValues.get(getVariable());
         if (value == null)
         {
-            final List<MissingValueException> missingValues = new ArrayList<MissingValueException>();
-            missingValues.add(new MissingValueException("Missing value for " + getVariable().getKey(),
-                    getVariable()));
-            throw new MissingValuesException("The calculation is missing values.", missingValues);
+            throw new MissingValuesException(ImmutableSet.of(getVariable()));
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/domain/model/SingleVariableTerm.java]: Overall, I like these changes a lot better than what it was doing previously.

File: srcalc/src/main/java/gov/va/med/srcalc/web/controller/EnterVariablesController.java [hunk 3]
```diff
@@ -135,18 +134,18 @@ public ModelAndView enterVariables(
...
                 LOGGER.debug("Field Error: {}", dynamicKey);
                 // Account for field errors on special fields like discrete numerical
                 // variables.
-                if(bindingResultAlreadyContainsError(dynamicKey, valuesBindingResult, missingValue))
+                if (!bindingResultAlreadyContainsError(missingValue, valuesBindingResult))
                 {
-                            valuesBindingResult.rejectValue(
-                                    dynamicKey, ERROR_NO_VALUE, missingValue.getMessage());
+                    valuesBindingResult.rejectValue(
+                            dynamicKey, ERROR_NO_VALUE, "no value");
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/web/controller/EnterVariablesController.java]: Should this be a constant somewhere?

File: srcalc/src/main/java/gov/va/med/srcalc/web/controller/EnterVariablesController.java [hunk 4]
```diff
@@ -164,12 +163,18 @@ public ModelAndView enterVariables(
...
+    private boolean bindingResultAlreadyContainsError(final Variable variable,
+            final BindingResult valuesBindingResult)
     {
+        final String baseKey = VariableEntry.makeDynamicValuePath(variable.getKey());
         final String numericalKey = VariableEntry.makeDynamicValuePath(
-                VariableEntry.makeNumericalInputName(missingValue.getVariable().getKey()));
-        return !valuesBindingResult.hasFieldErrors(dynamicKey) && !valuesBindingResult.hasFieldErrors(numericalKey);
+                VariableEntry.makeNumericalInputName(variable.getKey()));
+        return valuesBindingResult.hasFieldErrors(baseKey) ||
+                valuesBindingResult.hasFieldErrors(numericalKey);
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/web/controller/EnterVariablesController.java]: Good boolean logic change here.

File: srcalc/src/main/webapp/WEB-INF/views/displayResults.jsp [hunk 5]
```diff
@@ -27,13 +27,14 @@ Note: this view is a fusion of two different objects:
...
     </c:forEach>
     </table>
-    <div class="actionButtons">
     <h3>Results</h3>
-    <ol>
+    <!-- The outcomes are actually sorted by model name, but use a <ul> anyway. -->
+    <ul id="outcomesList">
     <c:forEach var="outcome" items="${result.outcomes}">
-    <li>${outcome.key}: <fmt:formatNumber value="${outcome.value * 100}" minFractionDigits="1" maxFractionDigits="1" />%</li>
+    <li>${outcome.key}: <fmt:formatNumber value="${outcome.value}" type="percent" minFractionDigits="1" maxFractionDigits="1" /></li>
```
Comment (jamesbigmac) [inline srcalc/src/main/webapp/WEB-INF/views/displayResults.jsp]: Didn't know you could do this.

File: srcalc/src/test/java/gov/va/med/srcalc/domain/model/DisplayNameComparatorTest.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package gov.va.med.srcalc.domain.model;
+
+import gov.va.med.srcalc.test.util.TestHelpers;
+
+import org.junit.Test;
+
+/**
+ * <p>Tests the {@link DisplayNameComparator} class.</p>
+ * 
+ * <p>Note that these test cases use BDD-style names "should {do something}".</p>
```
Comment (jamesbigmac) [inline srcalc/src/test/java/gov/va/med/srcalc/domain/model/DisplayNameComparatorTest.java]: I get the logic, but why the change here?

File: srcalc/src/main/java/gov/va/med/srcalc/domain/model/SingleVariableTerm.java (hunk 2 above)
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/domain/model/SingleVariableTerm.java]: Thanks!

File: srcalc/src/main/java/gov/va/med/srcalc/web/controller/EnterVariablesController.java (hunk 3 above)
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/web/controller/EnterVariablesController.java]: I don't think so: it's just a message string for humans, it has no meaning to the code.

File: srcalc/src/main/webapp/WEB-INF/views/displayResults.jsp (hunk 5 above)
Comment (dgtombs) [inline srcalc/src/main/webapp/WEB-INF/views/displayResults.jsp]: Me neither until a few weeks ago, which is why I wrote it in a more awkward way before.

File: srcalc/src/test/java/gov/va/med/srcalc/domain/model/DisplayNameComparatorTest.java (hunk 6 above)
Comment (dgtombs) [inline srcalc/src/test/java/gov/va/med/srcalc/domain/model/DisplayNameComparatorTest.java]: You mean the test names? I had already written a couple tests like that, and I think this naming style has many benefits. I highly recommend reading [Introducing BDD](http://dannorth.net/introducing-bdd/) if you haven't already.
